language/move-analyzer/src/symbols.rs

```diff
@@ -0,0 +1,2184 @@
+    col_end: u32,
+    /// Location of the definition
+    def_loc: DefLoc,
+}
+
+/// Definition of a struct field
+#[derive(Debug)]
+struct FieldDef {
+    name: Symbol,
+    start: Position,
+}
+
+/// Definition of a struct
+#[derive(Debug)]
+struct StructDef {
+    name_start: Position,
+    field_defs: Vec<FieldDef>,
+}
+
+/// Module-level definitions
+#[derive(Debug)]
+struct ModuleDefs {
+    /// File where this module is located
+    fhash: FileHash,
+    /// Struct definitions
+    structs: BTreeMap<Symbol, StructDef>,
+    /// Const definitions
+    constants: BTreeMap<Symbol, Position>,
+    /// Function definitions
+    functions: BTreeMap<Symbol, Position>,
+}
+
+/// Holds information about local definitions in a given scope
+type Scope = BTreeMap<Symbol, DefLoc>;
+
+/// Data used during symbolication
+pub struct Symbolicator {
+    /// Outermost definitions in a module (structs, consts, functions)
+    mod_outer_defs: BTreeMap<ModuleIdent_, ModuleDefs>,
+    /// A mapping from file names to file content (used to obtain source file locations)
+    files: SimpleFiles<Symbol, String>,
+    /// A mapping from file hashes to file IDs (used to obtain source file locations)
+    file_id_mapping: BTreeMap<FileHash, usize>,
+    /// Scope to contain type params where relevant (e.g. when processing function definition)
+    type_params: Scope,
+    /// Current processed module (always set before module processing starts)
+    current_mod: Option<ModuleIdent_>,
+}
+
+/// Result of the symbolication process
+pub struct Symbols {
+    /// A map from def locations to all the references (uses)
+    references: BTreeMap<DefLoc, BTreeSet<UseLoc>>,
+    /// A mapping from uses to definitions in a module
+    mod_use_defs: BTreeMap<ModuleIdent_, UseDefMap>,
+    /// A mapping from paths to module IDs
+    mod_ident_map: BTreeMap<PathBuf, ModuleIdent_>,
+    /// A mapping from file hashes to file names
+    file_name_mapping: BTreeMap<FileHash, Symbol>,
+}
+
+impl UseDef {
+    fn new(
+        references: &mut BTreeMap<DefLoc, BTreeSet<UseLoc>>,
+        use_fhash: FileHash,
+        use_start: Position,
+        def_fhash: FileHash,
+        def_start: Position,
+        use_name: &Symbol,
+    ) -> Self {
+        let def_loc = DefLoc {
+            fhash: def_fhash,
+            start: def_start,
+        };
+        let col_end = use_start.character + use_name.len() as u32;
+        let use_def = Self {
+            col_start: use_start.character,
+            col_end,
+            def_loc: def_loc.clone(),
+        };
+
+        let use_loc = UseLoc {
+            fhash: use_fhash,
+            start: use_start,
+            col_end,
+        };
+
+        match references.get_mut(&def_loc) {
+            Some(v) => {
+                v.insert(use_loc);
+            }
+            None => {
+                let mut s = BTreeSet::new();
+                s.insert(use_loc);
+                references.insert(def_loc, s);
+            }
+        }
```

Comment:
references.entry(def_loc).or_insert_with(BTreeSet::new).insert(use_loc)

Reply:
This is pretty cool!